gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/map/DenseIntValueHashMap.java

```diff
@@ -0,0 +1,353 @@
+
+  private static final byte[] TOMBSTONE = new byte[0];
+
+  protected final DenseHashMapSerializer<K> serializer;
+
+  private float loadFactor = 0.85f;
+
+  private final DenseIntValueHashMap.Segment<K>[] segments;
+
+  private static final boolean TRACE = true;
+
+  public DenseIntValueHashMap() {
+    this(new DHMDefaultSerializer(), 32, -1);
+  }
+
+  public DenseIntValueHashMap(int initialCapacity) {
+    this(new DHMDefaultSerializer(), initialCapacity, -1);
+  }
+
+  public DenseIntValueHashMap(int initialCapacity,
+      int concurrency) {
+    this(new DHMDefaultSerializer(), initialCapacity, concurrency);
+  }
+
+  public DenseIntValueHashMap(DenseHashMapSerializer serializer) {
+    this(serializer, 32, -1);
+  }
+
+  public DenseIntValueHashMap(DenseHashMapSerializer serializer,
+      int initialCapacity) {
+    this(serializer, initialCapacity, -1);
+  }
+
+  public DenseIntValueHashMap(DenseHashMapSerializer serializer,
+      int initialCapacity,
+      int concurrency) {
+
+    this.serializer = serializer;
+    if (concurrency <= 0) {
+      concurrency = Runtime.getRuntime().availableProcessors() * 8;
+    }
+    segments = new DenseIntValueHashMap.Segment[concurrency];
+
+    int capacity = 1;
+    // make capacity power of 2 instead of just taking a rounded off number.
+    // a rounded off number isn't good for probing.
+    while (capacity < initialCapacity)
+      capacity *= 2;
+
+    for (int i = 0; i < segments.length; i++) {
+      segments[i] = new DenseIntValueHashMap.Segment(i, serializer, loadFactor, capacity);
+    }
+  }
+
+  public final int size() {
+    int sz = 0;
+    for (Segment s : segments) {
+      if (s != null) {
+        sz += s.size;
+      }
+    }
+    return sz;
+  }
+
+  public void put(K key, int value) {
+    int hash = serializer.getHashCode(key);
+    segments[hash & (segments.length - 1)].putEntry(key, hash, value, false);
+  }
+
+  public void putIfAbsent(K key, int value) {
+    int hash = serializer.getHashCode(key);
+    segments[hash & (segments.length - 1)].putEntry(key, hash, value, true);
+  }
+
+  public int get(K key) {
+    int hash = serializer.getHashCode((K)key);
+    return segments[hash & (segments.length - 1)].getEntry(key, hash);
+  }
+
+  public int remove(K key) {
+    int hash = serializer.getHashCode((K)key);
+    return segments[hash & (segments.length - 1)].removeEntry(key, hash);
+  }
+
+  public final static int readInt(byte[] kv, int offset, int numBytes) {
+    return RowFormatter.readInt(kv, offset, numBytes);
+  }
+
+  public final static void writeInt(byte[] kv, int val, int offset, int numBytes) {
+    assert val < 0xffff || numBytes != 2;
+    RowFormatter.writeInt(kv, val, offset, numBytes);
+  }
+
+  static final class Segment<K> extends ReentrantLock {
+
+    private final int id;
+    private final DenseHashMapSerializer serializer;
+    private final float loadFactor;
+
+    private AtomicReferenceArray<byte[]> table;
```

Comment:
This should be volatile since rehash can replace it and get() reads it without a lock.